connmgr.go

```diff
@@ -305,6 +355,68 @@ func (cm *BasicConnMgr) trim() {
 	cm.lastTrimMu.Unlock()
 }
 
+func (cm *BasicConnMgr) getConnsToCloseEmergency(target int) []network.Conn {
+	npeers := cm.segments.countPeers()
+	candidates := make(peerInfos, 0, npeers)
+
+	cm.plk.RLock()
+	for _, s := range cm.segments {
+		s.Lock()
+		for id, inf := range s.peers {
+			if _, ok := cm.protected[id]; ok {
+				// skip over protected peer.
+				continue
+			}
+			candidates = append(candidates, inf)
+		}
+		s.Unlock()
+	}
+	cm.plk.RUnlock()
+
+	// Sort peers according to their value.
+	candidates.SortByValue()
+
+	selected := make([]network.Conn, 0, target+10)
+	for _, inf := range candidates {
+		if target <= 0 {
+			break
+		}
+		for c := range inf.conns {
+			selected = append(selected, c)
+		}
+		target -= len(inf.conns)
+	}
+	if len(selected) >= target {
+		// We found enough connections that were not protected.
+		return selected
+	}
+
+	// We didn't find enough unprotected connections.
+	// We have no choice but to kill some protected connections.
+	candidates = candidates[:0]
```

Comment:
this second loop is slightly problematic, because we may close too many protrcted conns.

I think it is better to just iterate again and append protected conns until we have enough instead of starting from scratch.

Reply:
ok this is fine.